# Linear Algebra: Manipulating Images with Matrix Operations
First, we load an image using the Python Image Processing Library. Note that even though we'll reimplement most of the library's functionality (crop, resize, scale, merge methods), we're not reimplementing image reading, as that would mean interpreting bytes based on the image spec (JPEG/PNG/whatever), and that isn't our focus.

In [2]:
import numpy as np
from PIL import Image

image = Image.open('fixtures/koala.jpg')

# This works because the Image class implements .__array_interface__()
matrix = np.array(image)
print(matrix.shape)
print(matrix[0][0])


(768, 1024, 3)
[101  90  58]


## Understanding the image
The image is 1024 x 768, in RGB mode. Its matrix has the shape (768, 1024, 3). So this is a 2-dimensional matrix (768 * 1024) with 3 items each, representing the RGB value of each pixel. For instance, printing `matrix[0]` gives `[101  90  58]`.

## Challenge 1: Size manipulation - Crop
Say we only want to keep the centre of the image, by area (ie we maintain the aspect ratio).

Area of image = 1024 * 768 = 786 432 pixels².
Desired area = 393 216 pixels²
To find the length and breadth to give us this aspect ratio: 4p * 3p = 12p² = 393216 => p = 181.02.

This gives us dimensions 4(181.02) x 3(181.02) = 724 x 543 (round to 544 to be an even number).

So, from the smaller dimension in the matrix (768), we will keep the inner 544 (remove 112 at the start and end). From the higher dimension (1024), we will keep the inner 724 (remove 150 at the start and end).

In [33]:
cropped = matrix[112:-112]
cropped = cropped[:,150:-150]

Image.fromarray(cropped).show()


## Challenge 2: Size manipulation - Resize
How do we reduce the size of this image (without changing its other dimensions)?
- We need to keep its original aspect ratio, which is 4:3 (1024:768)
- To reduce this, we must get rid of some pixels. We can't simply take the first/last 600. We must reduce proportionally.

Supposing we want to reduce it to 600x400. This is a reduction of 1.28. So we must replace every 1.28 pixels by 1 pixel, or every 64 pixels must be replaced by 50.

## Challenge 3: Geometrical manipulation - Rotate
Rotating an image by 90 is easy: we can transpose its pixels.

## Challenge 3: Geometrical manipulation - Reflect
To reflect geometrically around the y-axis, the mapping is f: x -> -x. Here, x is the horizontal coordinate (1024). So we only need to invert the indexes.

 

In [5]:
reflected_in_y = matrix[:,::-1]
# Image.fromarray(reflected_in_y).show()

reflected_in_x = matrix[::-1]
reflected_in_xy = matrix[::-1,::-1]
Image.fromarray(reflected_in_xy).show()


## Challenge 4: Colour manipulation - Filter

In [40]:
print(matrix[0, 0])
filtered = matrix
# We could also just do filtered[:,:,2] *= 0.5, but this fails because of the casting (float to int)
filtered[:,:,2] = np.multiply(0.5, filtered[:,:,2], casting='unsafe')
print(filtered[0, 0])

Image.fromarray(filtered).show()


[101  90  29]
[101  90  14]
